File: packages/mobile-react-native/android/src/main/java/com/rtcexpress/reactnative/RTCExpressModule.kt
package com.rtcexpress.reactnative

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.rtcexpress.sdk.CallInvite
import com.rtcexpress.sdk.CallStateUpdate
import com.rtcexpress.sdk.RTCExpress
import com.rtcexpress.sdk.RTCInitOptions
import com.rtcexpress.sdk.RoomMessage
import com.rtcexpress.sdk.TokenClient
import com.rtcexpress.sdk.TokenRequest

class RTCExpressModule(private val reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext), RTCExpress.Listener {

    private var rtc: RTCExpress? = null

    override fun getName() = "RTCExpress"

    private fun emit(event: String, payload: Any?) {
        reactContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit(event, payload)
    }

    @ReactMethod
    fun fetchToken(serverUrl: String, request: ReadableMap, promise: Promise) {
        try {
            val tokenRequest = TokenRequest(
                appId = request.getString("appId") ?: "",
                appSecret = request.getString("appSecret") ?: "",
                userId = request.getString("userId") ?: "",
                roomId = if (request.hasKey("roomId")) request.getString("roomId") else null
            )
            val response = TokenClient().fetchToken(serverUrl, tokenRequest)
            val map = Arguments.createMap()
            map.putString("token", response.token)
            map.putInt("expiresIn", response.expiresIn)
            promise.resolve(map)
        } catch (e: Exception) {
            promise.reject("token_error", e.message, e)
        }
    }

    @ReactMethod
    fun init(options: ReadableMap, promise: Promise) {
        try {
            val rtcExpress = RTCExpress(reactContext.applicationContext)
            rtcExpress.setListener(this)
            rtcExpress.init(
                RTCInitOptions(
                    serverUrl = options.getString("serverUrl") ?: "",
                    appId = options.getString("appId") ?: "",
                    userId = options.getString("userId") ?: "",
                    token = options.getString("token") ?: "",
                    mediaMode = options.getString("mediaMode") ?: "auto"
                )
            )
            rtc = rtcExpress
            promise.resolve(null)
        } catch (e: Exception) {
            promise.reject("init_error", e.message, e)
        }
    }

    @ReactMethod
    fun joinRoom(roomId: String) {
        rtc?.joinRoom(roomId)
    }

    @ReactMethod
    fun sendMessage(text: String) {
        rtc?.sendMessage(text)
    }

    @ReactMethod
    fun callUser(peerUserId: String, video: Boolean) {
        rtc?.callUser(peerUserId, video)
    }

    @ReactMethod
    fun acceptCall() {
        rtc?.acceptCall()
    }

    @ReactMethod
    fun rejectCall() {
        rtc?.rejectCall()
    }

    @ReactMethod
    fun endCall() {
        rtc?.endCall()
    }

    @ReactMethod
    fun muteMicrophone(muted: Boolean) {
        rtc?.muteMicrophone(muted)
    }

    @ReactMethod
    fun muteCamera(muted: Boolean) {
        rtc?.muteCamera(muted)
    }

    @ReactMethod
    fun switchCamera() {
        rtc?.switchCamera()
    }

    @ReactMethod
    fun destroy() {
        rtc?.destroy()
        rtc = null
    }

    override fun onConnected(userId: String) {
        emit("connected", Arguments.createMap().apply { putString("userId", userId) })
    }

    override fun onDisconnected() {
        emit("disconnected", null)
    }

    override fun onRoomJoined(roomId: String, members: List<String>) {
        emit(
            "roomJoined",
            Arguments.createMap().apply {
                putString("roomId", roomId)
                putArray("members", Arguments.fromList(members))
            }
        )
    }

    override fun onMessage(message: RoomMessage) {
        emit(
            "message",
            Arguments.createMap().apply {
                putString("roomId", message.roomId)
                putString("fromUserId", message.fromUserId)
                putString("text", message.text)
                putDouble("sentAt", message.sentAt.toDouble())
            }
        )
    }

    override fun onCallInvite(invite: CallInvite) {
        emit(
            "callInvite",
            Arguments.createMap().apply {
                putString("callId", invite.callId)
                putString("roomId", invite.roomId)
                putString("fromUserId", invite.fromUserId)
                putString("callType", invite.callType)
            }
        )
    }

    override fun onCallState(update: CallStateUpdate) {
        emit(
            "callState",
            Arguments.createMap().apply {
                putString("callId", update.callId)
                putString("state", update.state)
                putString("peerUserId", update.peerUserId)
                putString("roomId", update.roomId)
                putString("callType", update.callType)
            }
        )
    }

    override fun onRemoteVideo(track: org.webrtc.VideoTrack) {
        emit("remoteVideo", Arguments.createMap().apply { putString("trackId", track.id()) })
    }

    override fun onError(message: String) {
        emit("error", Arguments.createMap().apply { putString("message", message) })
    }
}

File: packages/mobile-react-native/ios/RTCExpressModule.swift
import Foundation
import React

@objc(RTCExpress)
class RTCExpressModule: RCTEventEmitter, RTCExpressDelegate {
    private var rtc: RTCExpress?

    override static func requiresMainQueueSetup() -> Bool { true }

    override func supportedEvents() -> [String]! {
        ["connected", "disconnected", "roomJoined", "message", "callInvite", "callState", "remoteVideo", "error"]
    }

    @objc(fetchToken:request:resolver:rejecter:)
    func fetchToken(
        _ serverUrl: String,
        request: NSDictionary,
        resolver resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        Task {
            do {
                let tokenRequest = TokenRequest(
                    appId: request["appId"] as? String ?? "",
                    appSecret: request["appSecret"] as? String ?? "",
                    userId: request["userId"] as? String ?? "",
                    roomId: request["roomId"] as? String
                )
                let response = try await TokenClient.fetchToken(serverUrl: serverUrl, request: tokenRequest)
                resolve(["token": response.token, "expiresIn": response.expiresIn])
            } catch {
                reject("token_error", error.localizedDescription, error)
            }
        }
    }

    @objc(init:resolver:rejecter:)
    func initialize(
        _ options: NSDictionary,
        resolver resolve: @escaping RCTPromiseResolveBlock,
        rejecter reject: @escaping RCTPromiseRejectBlock
    ) {
        let express = RTCExpress()
        express.delegate = self
        express.initClient(
            RTCInitOptions(
                serverUrl: options["serverUrl"] as? String ?? "",
                appId: options["appId"] as? String ?? "",
                userId: options["userId"] as? String ?? "",
                token: options["token"] as? String ?? "",
                mediaMode: options["mediaMode"] as? String ?? "auto"
            )
        )
        rtc = express
        resolve(nil)
    }

    @objc func joinRoom(_ roomId: String) { rtc?.joinRoom(roomId) }
    @objc func sendMessage(_ text: String) { try? rtc?.sendMessage(text) }
    @objc func callUser(_ peerUserId: String, video: Bool) { rtc?.callUser(peerUserId, video: video) }
    @objc func acceptCall() { try? rtc?.acceptCall() }
    @objc func rejectCall() { rtc?.rejectCall() }
    @objc func endCall() { rtc?.endCall() }
    @objc func muteMicrophone(_ muted: Bool) { rtc?.muteMicrophone(muted) }
    @objc func muteCamera(_ muted: Bool) { rtc?.muteCamera(muted) }
    @objc func switchCamera() { rtc?.switchCamera() }
    @objc func destroy() { rtc = nil }

    func rtcExpress(_ rtc: RTCExpress, didConnect userId: String) {
        sendEvent(withName: "connected", body: ["userId": userId])
    }

    func rtcExpressDidDisconnect(_ rtc: RTCExpress) {
        sendEvent(withName: "disconnected", body: nil)
    }

    func rtcExpress(_ rtc: RTCExpress, didJoinRoom roomId: String, members: [String]) {
        sendEvent(withName: "roomJoined", body: ["roomId": roomId, "members": members])
    }

    func rtcExpress(_ rtc: RTCExpress, didReceive message: RoomMessage) {
        sendEvent(withName: "message", body: [
            "roomId": message.roomId,
            "fromUserId": message.fromUserId,
            "text": message.text,
            "sentAt": message.sentAt
        ])
    }

    func rtcExpress(_ rtc: RTCExpress, didReceive invite: CallInvite) {
        sendEvent(withName: "callInvite", body: [
            "callId": invite.callId,
            "roomId": invite.roomId,
            "fromUserId": invite.fromUserId,
            "callType": invite.callType
        ])
    }

    func rtcExpress(_ rtc: RTCExpress, callState: CallStateUpdate) {
        sendEvent(withName: "callState", body: [
            "callId": callState.callId,
            "state": callState.state,
            "peerUserId": callState.peerUserId,
            "roomId": callState.roomId,
            "callType": callState.callType
        ])
    }

    func rtcExpress(_ rtc: RTCExpress, remoteVideoTrack: RTCVideoTrack) {
        sendEvent(withName: "remoteVideo", body: ["trackId": remoteVideoTrack.trackId])
    }

    func rtcExpress(_ rtc: RTCExpress, didError message: String) {
        sendEvent(withName: "error", body: ["message": message])
    }
}
